File: here-to-slay/Models/Game.cs
namespace HereToSlay.Models;

public enum GameState
{
    Waiting,
    InProgress,
    Finished
}

public class Game
{
    public string RoomCode { get; set; } = string.Empty;
    public List<Player> Players { get; set; } = new();
    public List<CardInstance> Deck { get; set; } = new();
    public List<CardInstance> MonsterDeck { get; set; } = new();
    public List<CardInstance> DiscardPile { get; set; } = new();
    public List<CardInstance> MonsterRow { get; set; } = new();
    public int MaxPlayers { get; set; } = 2;
    public string GameType { get; set; } = "classic";
    public int CurrentPlayerIndex { get; set; }
    public GameState State { get; set; } = GameState.Waiting;
    public int ActionPointsRemaining { get; set; }
    public string? WinnerId { get; set; }
    public string? LastMessage { get; set; }
    public int? LastRollDie1 { get; set; }
    public int? LastRollDie2 { get; set; }
    public int? LastRollModifier { get; set; }
    public int? LastRollTotal { get; set; }
}

File: here-to-slay-client/src/types/game.ts
export type GameState = "waiting" | "inProgress" | "finished";

export interface PartyRequirement {
  heroClass: string | null;
  count: number;
  genericHero: boolean;
}

export interface CardView {
  instanceId: string;
  definitionId: string;
  name: string;
  type: string;
  heroClass: string | null;
  rollThreshold: number;
  failIfRollAtOrBelow: number | null;
  failPenalty: string | null;
  imagePath: string;
  effectText: string;
  partyRequirements: PartyRequirement[];
  attachedItems: CardView[];
  isPartyLeader: boolean;
  partyLeaderAbility: string | null;
  partyLeaderAbilityValue: number | null;
  partyLeaderAbilityAltValue: number | null;
}

export interface PlayerView {
  id: string;
  name: string;
  type: string;
  partyLeader: CardView | null;
  hand: CardView[];
  party: CardView[];
  slainMonsters: CardView[];
  handCount: number;
  heroesRolledThisTurn: string[];
}

export interface PendingChoiceView {
  kind: string;
  prompt: string;
  minSelections: number;
  maxSelections: number;
  selectableCards: CardView[];
  options: string[];
  isYourChoice: boolean;
}

export interface GameView {
  roomCode: string;
  state: GameState;
  gameType: string;
  maxPlayers: number;
  currentPlayerIndex: number;
  actionPointsRemaining: number;
  winnerId: string | null;
  winnerName: string | null;
  lastMessage: string | null;
  lastRollDie1: number | null;
  lastRollDie2: number | null;
  lastRollModifier: number | null;
  lastRollTotal: number | null;
  deckCount: number;
  monsterDeckCount: number;
  discardCount: number;
  monsterRow: CardView[];
  players: PlayerView[];
  viewingPlayerId: string;
  pendingChoice: PendingChoiceView | null;
}

export interface SessionInfo {
  playerId: string;
  playerName: string;
  roomCode: string;
  isHost: boolean;
}

export const AP = {
  draw: 1,
  play: 1,
  heroRoll: 1,
  attack: 2,
  discardRedraw: 3,
} as const;
